crates/ruff_python_ast/src/source_code/locator.rs

```diff
@@ -153,116 +65,270 @@ impl<'a> Locator<'a> {
     }
 }
 
+/// Index for fast [Location] to byte offset conversions.
+#[derive(Debug, Clone)]
+enum Index {
+    /// Optimized index for an ASCII only document
+    Ascii(AsciiIndex),
+
+    /// Index for UTF8 documents
+    Utf8(Utf8Index),
+}
+
+impl Index {
+    /// Truncate a [`Location`] to a byte offset in source code.
+    fn byte_offset(&self, location: Location, contents: &str) -> usize {
+        match self {
+            Index::Ascii(ascii) => ascii.byte_offset(location, contents),
+            Index::Utf8(utf8) => utf8.byte_offset(location, contents),
+        }
+    }
+}
+
+impl From<&str> for Index {
+    fn from(contents: &str) -> Self {
+        assert!(u32::try_from(contents.len()).is_ok());
+
+        let mut line_start_offsets: Vec<u32> = Vec::with_capacity(48);
```

Comment:
(For context, 48 was chosen based on some rough benchmarking of various constants, so it's sort of arbitrary but not totally arbitrary.)